File: src/main/java/com/ggardet/core/enumeration/ValueOfEnumValidator.java
package com.ggardet.core.enumeration;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import java.util.List;
import java.util.Locale;
import java.util.Objects;
import static java.util.stream.Stream.of;

class ValueOfEnumValidator implements ConstraintValidator<ValueOfEnum, CharSequence> {
    private List<String> acceptedValues;

    @Override
    public void initialize(ValueOfEnum annotation) {
        acceptedValues = of(annotation.enumClass().getEnumConstants())
                .map(Enum::name)
                .toList();
    }

    @Override
    public boolean isValid(CharSequence value, ConstraintValidatorContext context) {
        if (Objects.isNull(value)) {
            return true;
        }
        return acceptedValues.contains(value.toString().toUpperCase(Locale.ROOT));
    }
}

File: src/main/kotlin/com/ggardet/core/enumeration/ValueOfEnumValidator.kt
package com.ggardet.core.enumeration

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import java.util.*

internal class ValueOfEnumValidator : ConstraintValidator<ValueOfEnum, CharSequence> {
    private lateinit var acceptedValues: List<String>

    override fun initialize(annotation: ValueOfEnum) {
        acceptedValues = annotation.enumClass.java.enumConstants
            .map { it.name }
            .toList()
    }

    override fun isValid(value: CharSequence?, context: ConstraintValidatorContext?): Boolean {
        if (value == null) {
            return true
        }
        return acceptedValues.contains(value.toString().uppercase(Locale.ROOT))
    }
}